### PR TITLE
sql: allow selecting from the output of SHOW

### PR DESCRIPTION
**Pull request by knz**
Two patches here:

### sql: allow selecting from the output of SHOW

Unix has "everything is a file"; likewise, one of the greatest
features of SQL is "if there's data, then there's SELECT".

Like Unix has exposed block devices, scanners, printers and all
manners of strange devices through a file interface for programs to
`read()` and `write()` to them, PostgreSQL with `pg_catalog` and
really most SQL RDBMs with `information_schema` have exposed their
metadata as a data source that can be SELECTed from.

The striking commonality between those two cases is to subsume things
of wildly different nature under a common mechanism, *even if at first
it may not have made sense*, because in the long turn generality
trumps special cases -- an argument used by E.W. Dijsktra in the
1970's to support the introduction of general-purpose recursion in
ALGOL, amidst cries from his contemporaries that this feature could
not be possibly useful to anyone and that it never could be
implemented efficiently.  (For more details about this breathtaking
phase of computing history, see
http://www.dijkstrascry.com/sites/default/files/papers/preprint_0.pdf)

CockroachDB already embraces the future and tips its metaphoric hat to
Dijkstra by making one step further, exposing the output of `EXPLAIN`
as a data source via the `[...]` syntax. For example, `SELECT
max(level) FROM [EXPLAIN SELECT ....]` is really a thing in
CockroachDB and computes the depth of the query plan right out of the
SQL engine.

And then, there's MySQL. Its `SHOW` commands can produce tons of data,
but there is no way to manipulate this data within the SQL language.
CockroachDB did adopt MySQL's `SHOW`, for no other reason than
historical context (early CockroachDB did take inspiration from MySQL
about several things) and the practical observation that many use
cases require a shorthand syntax for what may be otherwise a complex
query into some virtual tables in `information_schema`, `pg_catalog`,
`system` or `crdb_internal`. *However* doing so forever prevents a
user from manipulating the output of `SHOW` in useful ways.

Forever... Forever? No!

In a blazing strike of `go-yacc` and `structs` and directed code
growth, this patch comes to rescue CockroachDB from the 1970's (truth
be told; the 1970's were fine era culturally; if it were not for very
poor design choices in the computing industry) and usher it into
Dijkstra's arms.

```
SELECT ordinality as flowers, 'flower for ' || name || ', my love' as present
  FROM [SHOW ALL CLUSTER SETTINGS]
  WITH ORDINALITY;
----
18 rows
flowers present
1       flower for diagnostics.reporting.enabled, my love
2       flower for diagnostics.reporting.report_metrics, my love
3       flower for diagnostics.reporting.send_crash_reports, my love
4       flower for kv.allocator.lease_rebalancing_aggressiveness, my love
5       flower for kv.allocator.load_based_lease_rebalancing.enabled, my love
6       flower for kv.raft.command.max_size, my love
7       flower for kv.raft_log.synchronize, my love
8       flower for kv.snapshot_rebalance.max_rate, my love
9       flower for kv.snapshot_recovery.max_rate, my love
10      flower for kv.transaction.max_intents, my love
11      flower for server.remote_debugging.mode, my love
12      flower for sql.defaults.distsql, my love
13      flower for sql.metrics.statement_details.dump_to_logs, my love
14      flower for sql.metrics.statement_details.enabled, my love
15      flower for sql.metrics.statement_details.threshold, my love
16      flower for sql.trace.log_statement_execute, my love
17      flower for sql.trace.session_eventlog.enabled, my love
18      flower for sql.trace.txn.enable_threshold, my love
```

And you may call "Why?" like Dijkstra's detractors, and to you,
friend, I say: you shall wait for our users to tell us, and we shall
learn.

### cli/sql: add a new table display format "raw"

**tl;dr** use `\set display_format raw` before a query to get the strings unmodified

Background
----------

One of the inconveniences of SHOW CREATE TABLE is that its output, by
default, encloses the generated CREATE TABLE statement in an ASCII art
table, like this:

```
root@:26257/> SHOW CREATE TABLE t.kv;
+-------+------------------------------------------------+
| Table |                  CreateTable                   |
+-------+------------------------------------------------+
| t.kv  | CREATE TABLE kv (␤                             |
|       |     k INT NOT NULL,␤                           |
|       |     v INT NULL,␤                               |
|       |     CONSTRAINT "primary" PRIMARY KEY (k ASC),␤ |
|       |     FAMILY "primary" (k, v)␤                   |
|       | )                                              |
+-------+------------------------------------------------+
(1 row)
```

This prevents copy-pasting the output directly onto some other
shell or text file.

Meanwhile, our SQL CLI shell *does* provide alternative ways to print
out the results, that can be selected with `\set display_format ...`;
however, none of the formats defined so far are able to print out the
above CREATE TABLE statement **in a way that can be copy-pasted
elsewhere**; for example:

```
root@:26257/> \set display_format html
root@:26257/> SHOW CREATE TABLE t.kv;
<table>
<thead><tr><th>Table</th><th>CreateTable</th></tr></head>
<tbody>
<tr><td>t.kv</td><td>CREATE TABLE kv (<br/>     k INT NOT NULL,<br/>    v INT NULL,<br/>
    CONSTRAINT &#34;primary&#34; PRIMARY KEY (k ASC),<br/>  FAMILY &#34;primary&#34; (k, v)<br/>)</td></tr>
</tbody>
</table>
```

(This replaces all special characters)

```
root@:26257/> \set display_format csv
root@:26257/> SHOW CREATE TABLE t.kv;
1 row
Table,CreateTable
t.kv,"CREATE TABLE kv (
        k INT NOT NULL,
        v INT NULL,
        CONSTRAINT ""primary"" PRIMARY KEY (k ASC),
        FAMILY ""primary"" (k, v)
)"
```

(This escapes the double quotes by doubling them, which renders the SQL invalid)

And so forth.

Solution: new display format
----------------------------

The `cockroach sql` CLI shell (as well as any built-in command that
emits tables) can use a custom display format configurable either with
`--format` on the command-line, or `\set display_format` in the shell.

This patch adds a new format `raw` which causes the string
representation of every record to be printed as-is, starting on its
own line, so that it can be copy-pasted from a terminal.

This format makes an extra mile to ensure the data can be further
processed: each record is prefixed by an annotation of the form `# N`
followed by a newline character, where *N* is the number of characters
in the record, starting with the first character following the
newline. A subsequent automated processing tool can use this
information to determine record boundaries.

Application
-----------

```
root@:26257/> select createtable from [show create table t.kv];
...
CREATE TABLE kv (
        k INT NOT NULL,
        v INT NULL,
        CONSTRAINT "primary" PRIMARY KEY (k ASC),
        FAMILY "primary" (k, v)
)
...
```

Tada!

### CODE REVIEW

**Comment by petermattis** [comment]
I'm a fan of doing this, but not for 1.0. I imagine we'll start to see more such PRs. Let's tag them 1.1.

**Comment by spencerkimball** [comment]
Fantastic PR description!

**Comment by knz** [comment]
@andreimatei @irfansharif wanna review this now 1.0 is out?

**Comment by irfansharif** [comment]
:lgtm: but please know I have less than minimal context into `pkg/(sql|parser)`, would not know if there was something amiss (also, lovely PR description haha).

---

Reviewed 9 of 9 files at r1, 4 of 4 files at r2.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15590#-:-KkBDovnI_SVx3b51Gw-:b-hk0rm0)*

**Comment by knz** [comment]
Thanks for the vote of confidence :)
Perhaps David you can chime in too for the green light then, since you know about these packages more these days.

**Comment by eisenstatdavid** [comment]
Code looks good, but I'd like a little more test coverage -- SELECT 1 FROM [...] LIMIT 1 doesn't reveal much about what ... did.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/testdata/logic_test/show_source, line 117 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15590#-KkBa9WyEQR-T5LO9-w7:-KkBa9WyEQR-T5LO9-w8:b-orvr4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/33cc8066049c5029c4ae5702f22c11fcbfe068fb/pkg/sql/testdata/logic_test/show_source#L117)):*
> ```
> CREATE INDEX ix ON foo(x); SELECT 1 FROM [SHOW TESTING_RANGES FROM INDEX foo@ix] LIMIT 1
> ----
> 1
> ```

At least one of these tests should have an alias clause. Is there a sensible way to test WITH ORDINALITY?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15590#-:-KkBap1iaTeyJKRx45xw:b-p5r1rm)*

**Comment by knz** [comment]
would you feel comfortable with `SELECT * FROM [SHOW ...] LIMIT 0`? to check the schema?

And this with and without other alias clause things?
Perhaps some join thrown in?

(Suggestions welcome)

**Comment by eisenstatdavid** [comment]
Perhaps `SELECT t.client_encoding FROM [SHOW client_encoding] AS t` -> `UTF8`, since that should be a very stable setting.

**Comment by knz** [comment]
Ah, I see thanks.

**Comment by knz** [comment]
Actually I realized (by trying it out) that all of the outputs in that test are suitably stable. So I just used that for simplicity. PTAL

**Comment by eisenstatdavid** [comment]
:lgtm:

---

Reviewed 7 of 9 files at r1, 2 of 6 files at r3, 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15590#-:-KkGP-_mgYwOpy_x8gLO:bnfp4nl)*
